debugger/src/main/java/org/codice/acdebugger/api/PermissionUtil.java

```diff
@@ -0,0 +1,387 @@
+    final String actions =
+        debug
+            .reflection()
+            .invoke(
+                permission, "getActions", ReflectionUtil.METHOD_SIGNATURE_NO_ARGS_STRING_RESULT);
+
+    if (debug.reflection().isInstance("Lorg/osgi/framework/ServicePermission;", permission)) {
+      // for ServicePermission, we typically get something like "(service.id=14)"
+      // for the permission name which is not useful; so we first try to get the set
+      // of object classes representing the services and if we get a set then we can
+      // create permissions for each names. If we get null then we typically can rely
+      // on the permission name
+      final String[] objectClass =
+          debug.reflection().get(permission, "objectClass", "[Ljava/lang/String;");
+
+      if (objectClass != null) {
+        return Stream.of(objectClass)
+            .map(n -> getPermissionString(obj.name(), n, actions))
+            .collect(Collectors.toSet());
+      }
+    }
+    return Collections.singleton(
+        getPermissionString(
+            obj.name(),
+            debug
+                .reflection()
+                .invoke(
+                    permission, "getName", ReflectionUtil.METHOD_SIGNATURE_NO_ARGS_STRING_RESULT),
+            actions));
+  }
+
+  /**
+   * Finds missing service permissions for the given bundle/domain that corresponds to a 'GET'
+   * service permission referenced in the given service event.
+   *
+   * @param bundle the name of the bundle associated with the given domain
+   * @param domain the domain to check service permissions against
+   * @param serviceEvent the service event which we need to check permissions for
+   * @return a set of permission strings for all missing service permissions
+   */
+  @SuppressWarnings({
+    "squid:S1181", /* letting VirtualMachineErrors bubble out directly, so ok to catch Throwable */
+    "squid:S1148" /* this is a console application */
+  })
+  public Set<String> findMissingServicePermissionStrings(
+      String bundle, ObjectReference domain, ObjectReference serviceEvent) {
+    try { // first try via the backdoor
+      final ServicePermissionInfo info =
+          debug
+              .backdoor() // only grant in continuous mode and if granting
+              .getServicePermissionInfoAndGrant(
+                  debug, bundle, domain, serviceEvent, debug.isContinuous() && debug.isGranting());
+
+      final Set<String> permissionStrings = info.getPermissionStrings();
+
+      if (debug.isContinuous()) {
+        // first make sure we cache all permissions since we granted them if they were missing
+        permissionStrings.forEach(p -> debug.getContext().grantPermission(bundle, p));
+      }
+      // next cache all info we have about pre-existed implied permissions
+      // and remove them from the complete list of permissions since they were not missing
+      info.getImpliedPermissionStrings()
+          .stream()
+          .peek(permissionStrings::remove)
```

Comment:
✏️ I still believe this is a violation of the intent of `peek` but I am not going to block this PR over a single line.